pkg/cloudprovider/providers/gce/cloud/mock/mock.go

```diff
@@ -226,3 +227,168 @@ func InsertAlphaAddressHook(ctx context.Context, key *meta.Key, obj *alpha.Addre
 	projectID := m.ProjectRouter.ProjectID(ctx, meta.VersionBeta, "addresses")
 	return convertAndInsertAlphaAddress(key, obj, m.Objects, meta.VersionAlpha, projectID)
 }
+
+// Map from InstanceGroup key to list of Instances
+type instanceGroupAttributes struct {
+	instanceMap map[meta.Key][]*ga.InstanceWithNamedPorts
+}
```

Comment:
Per our offline discussion, we may want to add a mutex here in case parallelization is added to ingress or the service controller.

Also see how feasible it is to move some of the repeated hook logic to methods with `instanceGroupAttributes` as the receiver.